### PR TITLE
194402 - create s100 ess endpoint productnames

### PR DESCRIPTION
**Pull request by mayur1501364**
This PBI consists changes of [PBI : 194402](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/194402)

- Added code changes
- Added unit test cases

### CODE REVIEW

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductNamesValidatorTests.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            {
+                ProductNames = productIdentifiers,
+                CallbackUri = callbackUri
+            };
+            var result = _validator.TestValidate(model);
+            Assert.That(result.Errors.Count == 0);
+        }
+
+        [Test]
+        public void WhenEmptyProductIdentifiersInProductIdentifiersRequest_ThenReturnBadRequest()
```
Comment (shirinbt14926) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductNamesValidatorTests.cs:44]:
```suggestion
        public void WhenEmptyProductNamesInProductNamesRequest_ThenReturnBadRequest()
```

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductNamesValidatorTests.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                ProductNames = productIdentifiers,
+                CallbackUri = callbackUri
+            };
+            var result = _validator.TestValidate(model);
+            result.ShouldHaveValidationErrorFor(fb => fb.ProductNames);
+            Assert.That(result.Errors.Any(x => x.ErrorMessage == "productIdentifiers cannot be null or empty."));
+        }
+
+        [Test]
+        public void WhenZeroLengthProductIdentifiersInProductIdentifiersRequest_ThenReturnBadRequest()
```
Comment (shirinbt14926) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductNamesValidatorTests.cs:74]:
```suggestion
        public void WhenZeroLengthProductNamesInProductNamesRequest_ThenReturnBadRequest()
```
Similarly in other test cases as well

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductNamesValidatorTests.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
+﻿using System;
+using System.Linq;
+using FluentValidation.TestHelper;
+using NUnit.Framework;
+using UKHO.ExchangeSetService.API.Validation.V2;
+using UKHO.ExchangeSetService.Common.Models.V2.Request;
+
+namespace UKHO.ExchangeSetService.API.UnitTests.Validation
```
Comment (shirinbt14926) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductNamesValidatorTests.cs:8]: Can we create a subfolder V2 in Validation for Unit Tests as well, to keep it same as in API folder structure

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Services/ExchangeSetService.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                List<Error> errors;
+
+                if (validationResult.HasBadRequestErrors(out errors))
+                {
+                    _logger.LogError(EventIds.InvalidProductNames.ToEventId(), "Product name validation failed. | X-Correlation-ID : {correlationId}", correlationId);
+                    return ServiceResponseResult<ExchangeSetResponse>.BadRequest(new ErrorDescription { CorrelationId = correlationId, Errors = errors });
+                }
+            }
+
+            _logger.LogInformation(EventIds.CreateProductDataByProductNamesCompleted.ToEventId(), "Creation of Product data completed | X-Correlation-ID : {correlationId}", correlationId);
```
Comment (shirinbt14926) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Services/ExchangeSetService.cs:69]:
```suggestion
            _logger.LogInformation(EventIds.CreateProductDataByProductNamesCompleted.ToEventId(), "Creation of Product data for product Names completed | X-Correlation-ID : {correlationId}", correlationId);
```

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Logging/EventIds.cs [hunk 5]
```diff
@@ -856,8 +856,31 @@ public enum EventIds
...
+        /// <summary>
+        /// 805212 - Request for exchange set service product names endpoint is started.
+        /// </summary>
+        ESSPostProductNamesRequestStart = 805212,
+        /// <summary>
+        /// 805213 - Request for exchange set service product names endpoint is completed.
+        /// </summary>
+        ESSPostProductNamesRequestCompleted = 805213,
+        /// <summary>
+        /// 805214 - Request for exchange set service call started.
```
Comment (shirinbt14926) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Logging/EventIds.cs:869]:
```suggestion
        /// 805214 - Request for exchange set service call by product names started.
```
similar change in 805215

**Review by shirinbt14926 — COMMENTED** [review]
Minor comments please

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductNamesValidatorTests.cs (hunk 1 above)
Comment (mayur1501364) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductNamesValidatorTests.cs:44]: Changed method names

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductNamesValidatorTests.cs (hunk 2 above)
Comment (mayur1501364) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductNamesValidatorTests.cs:74]: Changed name

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductNamesValidatorTests.cs (hunk 3 above)
Comment (mayur1501364) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductNamesValidatorTests.cs:8]: Moved to subfolder

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Services/ExchangeSetService.cs (hunk 4 above)
Comment (mayur1501364) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Services/ExchangeSetService.cs]: Changed log

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Logging/EventIds.cs (hunk 5 above)
Comment (mayur1501364) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Logging/EventIds.cs:869]: Changed summary

**Review by Nadeem14456 — APPROVED** [review]
LGTM.

**Review by shirinbt14926 — APPROVED** [review]
LGTM. Thanks

**Review by jrippington — APPROVED** [review]
Looks good, thanks!

**Pull request merged**